File: JavaScript/Interview_problems/Task_Schedular_demo.js
/*

FR:
    Schedule a task for a future time
    Execute task at a scheduled time
    Support task priority
    Support recurring task
    cancel a scheduled task
    Thread safe scheduling
    Task should be executed async

*/

class Task {
  execute() {
    throw new Error("execute() must be implemented");
  }
}

class EmailTask extends Task {
  constructor(email = null) {
    super();
    this.email = email;
  }

  execute() {
  }
}

class PaymentTask extends Task {
  execute() {
  }
}

class ScheduledTask {
  constructor(id, task, executionTime, priority) {
    this.id = id;
    this.task = task;
    this.executionTime = executionTime;
    this.priority = priority;
  }
}

class TaskQueue {
  constructor() {
    this.queue = [];
  }

  add(task) {
    this.queue.push(task);
    this.queue.sort((a, b) => a.executionTime - b.executionTime || a.priority - b.priority);
  }

  peek() {
    return this.queue.length > 0 ? this.queue[0] : null;
  }
}

class TaskScheduler {
  constructor(taskQueue) {
    this.taskQueue = taskQueue;
  }

  schedule(task) {
    this.taskQueue.add(task);
  }
}

class Dispatcher {
  constructor(taskQueue) {
    this.taskQueue = taskQueue;
    this.isRunning = true;
  }

  run() {
    while (this.isRunning) {
      const task = this.taskQueue.peek();

      // what is the current time
      // delay =
    }
  }
}

class TaskSchedularDemo {
}

function main() {
}

main();

File: C++/Interview_problems/Task_Schedular_demo.cpp
/*

FR:
    Schedule a task for a future time
    Execute task at a scheduled time
    Support task priority
    Support recurring task
    cancel a scheduled task
    Thread safe scheduling
    Task should be executed async

*/

#include <string>
#include <queue>
#include <mutex>
#include <vector>
#include <memory>

struct Task {
    virtual void execute() = 0;
    virtual ~Task() = default;
};

class EmailTask : public Task {
public:
    std::string email;

    void execute() override {
    }
};

class PaymentTask : public Task {
public:
    void execute() override {
    }
};

class ScheduledTask {
public:
    std::string id;
    std::shared_ptr<Task> task;
    long executionTime;
    int priority;

    bool operator<(const ScheduledTask& other) const {
        if (executionTime != other.executionTime) {
            return executionTime > other.executionTime;
        }
        return priority > other.priority;
    }
};

class TaskQueue {
public:
    std::priority_queue<ScheduledTask> queue;
    std::mutex mtx;

    void add(const ScheduledTask& task) {
        std::lock_guard<std::mutex> lock(mtx);
        queue.push(task);
    }
};

class TaskScheduler {
public:
    std::shared_ptr<TaskQueue> taskQueue;

    explicit TaskScheduler(std::shared_ptr<TaskQueue> taskQueue) : taskQueue(taskQueue) {}

    void schedule(const ScheduledTask& task) {
        taskQueue->add(task);
    }
};

class Dispatcher {
public:
    std::shared_ptr<TaskQueue> taskQueue;
    bool isRunning = true;

    void run() {
        while (isRunning) {
            // what is the current time
            // delay =
        }
    }
};

class TaskSchedularDemo {
};

int main() {
    return 0;
}
